lib/state/wallet.ts

```diff
@@ -0,0 +1,340 @@
+];
+
+let accountsSubscriptionUnsub: VoidFunction | null = null;
+
+/**
+ * Enable a wallet by enabling the extension and setting the wallet atom state to connected.
+ * Also starts subscribing to accounts on the extension and updates the accounts in state.
+ *
+ * @param walletId the id or wallet itself to enable
+ * @returns Promise<boolean> - whether the wallet was enabled
+ */
+const enableWallet = async (walletId: Wallet | string) => {
+  if (accountsSubscriptionUnsub) accountsSubscriptionUnsub();
+
+  const wallet = isString(walletId)
+    ? supportedWallets.find((w) => w.extensionName === walletId)
+    : walletId;
+
+  try {
+    store.set(walletAtom, (state) => {
+      return {
+        ...state,
+        connected: false,
+      };
+    });
+
+    await wallet.enable();
+
+    store.set(walletAtom, (state) => {
+      return {
+        ...state,
+        wallet,
+      };
+    });
+
+    accountsSubscriptionUnsub = await wallet.subscribeAccounts((accounts) => {
+      if (accounts.length === 0) {
+        store.set(walletAtom, (state) => {
+          return {
+            ...state,
+            accounts: [],
+            errorMessages: [
+              {
+                extensionName: wallet.extensionName,
+                message:
+                  "No accounts on this wallet. Please add account in wallet extension.",
+              },
+            ],
+          };
+        });
+      } else {
+        store.set(walletAtom, (state) => {
+          return {
+            ...state,
+            connected: true,
+            accounts: accounts.map((account) => {
+              return {
+                ...account,
+                address: encodeAddress(account.address, 73),
+              };
+            }),
+            errorMessages: [],
+          };
+        });
+      }
+    });
+
+    return true;
+  } catch (error) {
+    store.set(walletAtom, (state) => {
+      return {
+        ...state,
+        accounts: [],
+        errorMessages: [
+          {
+            extensionName: wallet.extensionName,
+            message:
+              "Not allowed to interact with extension. Please change permission settings.",
+          },
+        ],
+      };
+    });
+    return false;
+  }
+};
+
+/**
+ * Enable wallet on first load if wallet id is set.
+ */
+if (store.get(userConfigAtom).walletId) {
+  enableWallet(store.get(userConfigAtom).walletId);
+}
+
+/**
+ * Enable wallet when the selected wallet id changes.
+ */
+store.sub(userConfigAtom, () => {
+  const { walletId } = store.get(userConfigAtom);
+  if (walletId) {
+    enableWallet(walletId);
+  }
+});
+
+/**
+ * Subscribe to user location and vpn changes and
+ * disconnect the wallet if location isnt allowed or vpn is used.
+ */
+store.sub(userLocationDataAtom, async () => {
+  const data = await store.get(userLocationDataAtom);
+  if (
+    (store.get(walletAtom).connected && !data?.locationAllowed) ||
+    data?.isUsingVPN
+  ) {
+    store.set(walletAtom, (state) => {
```

Comment:
Might be nice to use `disconnectWallet` if it's doing the same thing

Reply:
`disconnectWallet` is only usable by the hook, and if I create one outside the hook that uses the store and then use it in the hook there are issues with state synchronization. I can explain on call, but will leave this for now.